Genbank error checking

Koeng101 opened the pull request:
This is a genbank error checking file. As with the other branches, multi genbank parsing was removed. Parse, write, and read now defaults to outputting a list, though we don't have multi-genbank working quite yet.

I deleted `TestLocusParseRegression` because _it doesn't work_. These two lines are different, which the old parser **did not pick up**, but the newer parser picks up.
https://github.com/TimothyStiles/poly/blob/359da2c15a5b20c12b68d5eb25f8254dd67ee62a/data/puc19static.gbk#L43

https://github.com/TimothyStiles/poly/blob/359da2c15a5b20c12b68d5eb25f8254dd67ee62a/data/puc19.gbk#L43

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -821,7 +785,7 @@ func buildMetaString(name string, data string) string {
 			returnData = name + datum + "\n"
 		} else {
-			returnData += generateWhiteSpace(11) + datum + "\n"
+			returnData += generateWhiteSpace(12) + datum + "\n"
```
why the change to 12?

Koeng101 replied:
Previously this behavior was just broken. See:

https://github.com/TimothyStiles/poly/blob/7f133c87716de08b873b44d5a405688da1ddebb4/data/puc19static.gbk#L10-L11

Koeng101 replied:
Should be aligned

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -803,6 +766,7 @@ func parseLocation(locationString string) Location {
 		location = location.SubLocations[0]
 	}
 
+	location.GbkLocationString = locationString
```
This assignment doesn't do anything?

Koeng101 replied:
Correct, removed. Sorry about that.

TimothyStiles commented:
Biggest concern is that the original Genbank/MultiGenbank paradigm has been overwritten where now MultiGenbank is now just being called Genbank. Can we have both?

Koeng101 commented on io/genbank/genbank.go:
```diff
@@ -72,7 +71,6 @@ type Feature struct {
 
 // Reference holds information for one reference in a Meta struct.
 type Reference struct {
-	Index   string `json:"index"`
```
Index was redundant I believe, literally coded twice.

Koeng101 commented:
> Biggest concern is that the original Genbank/MultiGenbank paradigm has been overwritten where now MultiGenbank is now just being called Genbank. Can we have both?

Yes - Genbank is just MultiGenbank of list size 1 (in fact, the parser only does single Genbank right now, gotta fix it up to do multi. Lemme try that out). Much easier than having separate functions I think. 

Removed redundancy in latest push and deleted puc19static - which was broken in multiple ways.

TimothyStiles commented:
@Koeng101 don't forget to merge upstream. I'm pretty sure I deleted that static puc19 file in my last merged PR.

TimothyStiles commented:
> > Biggest concern is that the original Genbank/MultiGenbank paradigm has been overwritten where now MultiGenbank is now just being called Genbank. Can we have both?
> 
> Yes - Genbank is just MultiGenbank of list size 1 (in fact, the parser only does single Genbank right now, gotta fix it up to do multi. Lemme try that out). Much easier than having separate functions I think.

@Koeng101 I think separate functions would be better for clarity but we can debate later. It'd be super easy to add a variadic argument to a multigenbank parser to limit number of genbanks parsed and just call that from our "single" genbank parser.

Koeng101 commented:
> @Koeng101 don't forget to merge upstream. I'm pretty sure I deleted that static puc19 file in my last merged PR.

Yes, it is merged upstream, and no, it wasn't deleted because it was used for testing.

> I think separate functions would be better for clarity but we can debate later

Ye that is fair

Koeng101 commented:
@TimothyStiles How is this looking? I recall I removed the multi-genbank functions, but is there anything else?

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -147,221 +145,290 @@ func getFeatureSequence(feature Feature, location Location) (string, error) {
-	gbkString.WriteString("FEATURES             Location/Qualifiers\n")
-	for _, feature := range sequence.Features {
-		gbkString.WriteString(BuildFeatureString(feature))
+func parseReferences(metadataData []string) (Reference, error) {
```
will remain to make it easier to batch rename variables when compared to metadataTag.

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -147,221 +145,290 @@ func getFeatureSequence(feature Feature, location Location) (string, error) {
+	case "REMARK":
+		reference.Remark = referenceValue
+	default:
+		return fmt.Errorf("ReferenceKey not in [AUTHORS, TITLE, JOURNAL, PUBMED, REMARK]. Got: %s", referenceKey)
```
any test case for this?

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -898,114 +861,3 @@ func generateWhiteSpace(length int) string {
 ******************************************************************************/
-
-/******************************************************************************
-
```
@Koeng101 did you mean to deprecate the below section?

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -146,222 +144,314 @@ func getFeatureSequence(feature Feature, location Location) (string, error) {
 
-	versionString := buildMetaString("VERSION", sequence.Meta.Version)
-	gbkString.WriteString(versionString)
+	genbankStarted := false
```
@Koeng101 was this supposed to be used somewhere? The flag is set to true but it doesn't seem to ever be checked.

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -146,222 +144,548 @@ func getFeatureSequence(feature Feature, location Location) (string, error) {
+		for i, taxonomyData := range sequence.Meta.Taxonomy {
+			taxonomyString.WriteString(taxonomyData)
+			if len(taxonomyData) == i+1 {
+				taxonomyString.WriteString(".")
```
@koeng this logic is covered in tests. What does it do exactly?

TimothyStiles commented on io/genbank/genbank.go:
```diff
@@ -146,222 +144,551 @@ func getFeatureSequence(feature Feature, location Location) (string, error) {
+					continue
+				}
+				// If the trimmed line ends, append to the attribute and set parameters.quoteActive set to false
+				if trimmedLine[len(trimmedLine)-1] == '"' {
```
this won't work with all labels. Labels in t4_intron.gb don't have quotes surrounding label string.

TimothyStiles replied:
attention @Koeng101 can you work on this bug? Essentially you can't rely on quotations to show you the end of an attribute because some one line attributes don't use any quotations like this:

```
 gene            1..243
                 /gene="frd.1"
                 /locus_tag="T4p240"
                 /label=frd.1  <--- offending line
                 /db_xref="GeneID:1258741"

```

TimothyStiles commented:
@Koeng101 found an issue with the genbank parser where it somehow strips out complements from the gbk string while parsing. Any idea on what's going on here?
![Screen Shot 2022-06-18 at 4 46 51 AM](https://user-images.githubusercontent.com/7042260/174436145-e5fc2084-bce9-443a-8e3b-82d473e2e9ea.png)
![Screen Shot 2022-06-18 at 4 45 58 AM](https://user-images.githubusercontent.com/7042260/174436161-2be43674-2cb7-4470-878c-1b2115a95754.png)

TimothyStiles commented:
Also need to fix qualifier line breaks but that should be easy enoug
![Screen Shot 2022-06-18 at 4 51 23 AM](https://user-images.githubusercontent.com/7042260/174436202-8202aa67-36d2-4dc6-8107-1aa138c5fdf7.png)
h
